crates/sui-adapter/src/adapter.rs

```diff
@@ -1353,7 +1390,31 @@ fn convert_type_argument_error(idx: usize, error: VMError) -> ExecutionError {
         StatusCode::TYPE_RESOLUTION_FAILURE => EntryTypeArgumentErrorKind::TypeNotFound,
         StatusCode::NUMBER_OF_TYPE_ARGUMENTS_MISMATCH => EntryTypeArgumentErrorKind::ArityMismatch,
         StatusCode::CONSTRAINT_NOT_SATISFIED => EntryTypeArgumentErrorKind::ConstraintNotSatisfied,
-        _ => return error.into(),
+        _ => return convert_vm_error(error, empty_resolver()),
     };
     ExecutionErrorKind::entry_type_argument_error(idx as TypeParameterIndex, kind).into()
 }
+
+pub fn empty_resolver() -> impl Fn(&ModuleId, FunctionDefinitionIndex) -> Option<Identifier> {
+    |_: &ModuleId, _: FunctionDefinitionIndex| None
+}
+
+pub fn with_vm_resolver<
+    'r,
+    E: Debug,
+    S: ResourceResolver<Error = E> + ModuleResolver<Error = E>,
+>(
+    vm: &'r MoveVM,
+    state_view: &'r S,
+) -> impl Fn(&ModuleId, FunctionDefinitionIndex) -> Option<Identifier> + 'r {
```

Comment:
![image](https://user-images.githubusercontent.com/888624/212436407-cd812a8e-1ec8-4fc6-9ea2-0261bfaeee6d.png)

Gonna be real, I just wanted to a helper function to create the resolver closure here. So re: lifetime annotations I just kind of tried things until the compiler was happy

Reply:
I would say Rust does not do so well with this functional style, I would suggest using traits:

```rust
trait FunctionNameResolver {
    fn resolve_fname(&self, m: &ModuleId, f: FunctionDefinitionIndex) ->  Option<Identifier>;
    fn convert_vm_error(&self, error: VMError) -> ExecutionError {
        /* ... */
    }
}

struct NullResolver;
impl FunctionNameResolver for NullResolver {
    fn resolve_fname(&self, _: &ModuleId, _: FunctionDefinitionIndex) -> Option<Identifier> {
        None
    }
}

struct VMResolver<'vm, E> {
    vm: &'vm MoveVM,
    state_view: &'vm impl MoveResolver,
}

impl FunctionNameResolver for VMResolver {
    fn resolve_fname(&self, m: &ModuleId, f: FunctionDefinitionIndex) -> Option<Identifier> {
        let module = vm.load_module(module_id, state_view).ok()?;
        let fdef = module.function_def_at(idx);
        let fhandle = module.function_handle_at(fdef.function);
        let fname = module.identifier_at(fhandle.name);
        Some(fname.to_owned())        
    }
}
```

Couple of other changes that I folded into this one:

- There's a convenient alias for the trait constraint on `state_view`: `MoveResolver`.
- I used `Result::ok` to convert it to an `Option` and then `?` syntax rather than `if let Ok(...) = ... { ...; Some(_) } else { None }`.

Reply:
This was super instructive, thank you! 

> There's a convenient alias for the trait constraint on state_view: MoveResolver.

Nice!

> I used Result::ok to convert it to an Option and then ? syntax rather than if let Ok(...) = ... { ...; Some(_) } else { None }

💯 

Some thoughts to tie up loose ends from my side:

- I played with the syntax a bit to get things to compile, ran into a couple things like not being able to do `impl MoveResolver` on the `state_view` value in the struct

- About `convert_vm_error`, I _think_ we want to keep this as a shared function in `error.rs` based on how things depend on it right now, and not a function that's part of the `FunctionNameResolver` trait. (It seems a bit strange to have something called `FunctionNameResolver` actually do error conversion and not just function resolving. Let me know if you had something else in mind here though).

Reply:
> About `convert_vm_error`, I think we want to keep this as a shared function in error.rs
 
I think this is fair!  No complaints from me on that.


> Not being able to do `impl MoveResolver` on the `state_view` in the struct.

This is either because the trait is not in scope (i.e. needs to be imported) or because Rust straight up doesn't support the impl syntax on structs.  I kind of suspect the latter?  In which case I *think* what you want to do here is to make the struct generic on `state_view`'s type and unconstrained, and then apply the constraint on your trait impl, something like:

```rust
struct VMResolver<'vm, S> {
    vm: &'vm MoveVM,
    state_view: &'vm S,
}

impl <'vm, S: MoveResolver> FunctionNameResolver for VMResolver<'vm, S> {
    /* ... */
} 
```

Reply:
Indeed, the latter! Your snippet is what I ended up arriving at 👍. Interestingly I think the compiler said it's an unstable/nightly feature.